Replace some "assert_throws(new FooError(), stuff)" calls with assert_throws_js

stephenmcgruer opened the pull request:
This is an attempt to understand what part of #21350 is failing the CI; please excuse the email spam when wpt-pr-bot cc's you to this PR.

bzbarsky committed:
Replace some "assert_throws(new FooError(), stuff)" calls with assert_throws_js.

This diff was generated by running:

  find . -type f -print0 | xargs -0 perl -pi -e 'BEGIN { $/ = undef; } s/assert_throws\(([ \n]*)new ([A-Za-z]*Error) *\(\) *(, *.)/assert_throws_js(\1\2\3/gs'

and then:

1) Manually adjusting fullscreen/rendering/fullscreen-pseudo-class-support.html
to test for the right sort of exceptions ("SyntaxError" DOMException, not a JS
SyntaxError).

2) Manually adjusting performance-timeline/po-observe-type.any.js to test for
the right sort of exceptions ("SyntaxError" DOMException, not a JS
SyntaxError).

3) Manually adjusting performance-timeline/po-observe.any.js to test for
the right sort of exceptions ("SyntaxError" DOMException, not a JS
SyntaxError).

4) Manually adjusting user-timing/mark_exceptions.html to test for the right
sort of exceptions ("SyntaxError" DOMException, not a JS SyntaxError).

5) Manually adjusting user-timing/measure_syntax_err.any.js to test for the right
sort of exceptions ("SyntaxError" DOMException, not a JS SyntaxError).

6) Manually adjusting domxpath/lexical-structure.html to test for a
"SyntaxError" DOMException, since that's what all browsers throw and there is no
clear spec for this.

7) Manually adjusting workers/constructors/Worker/Worker-constructor.html to
test for the right sort of exceptions ("SyntaxError" DOMException, not a JS
SyntaxError).

8) Backing out the changes to resources/idlharness.js because some tests pass
objects from a different window to it, and we end up with the wrong TypeError
constructor in those cases.

This does affect indentation poorly in cases when the first arg was on the same
line as the assert_throws, there was a newline after the ',' after the first
arg, and the following args were lined up with the first arg.  Fixing that,
especially when there are multiple lines after the first arg, is not trivial
with a regexp.

Co-authored-by: Boris Zbarsky <bzbarsky@mit.edu>
Co-authored-by: Stephen McGruer <smcgruer@chromium.org>

stephenmcgruer commented:
Ah, I've figured out what was happening here @bzbarsky . You were (unwittingly) updating tests that are commit-time generated, without changing the generation logic. So we run a CI to check that they don't change, and re-running the generation undid your work and left the CI check with a diff (which fails it). I'll figure out how to update the generation logic.

bzbarsky commented:
Aha!  I knew we had some pre-generated tests, but not ones that are commit-time generated....

bzbarsky commented on 2dcontext/tools/gentestutils.py:
```diff
@@ -233,7 +233,7 @@ def expand_test_code(code):
 
         code = re.sub(r'@assert throws (\S+Error) (.*);',
-                r'assert_throws(new \1(), function() { \2; });',
+                r'assert_throws_js(\1, function() { \2; });',
```
The bit right above this probably needs to become `assert_throws_dom("\1", etc`, right?

stephenmcgruer replied:
Yes, I will do that when I get to the commit where you did `assert_throws_dom` rewrites.

stephenmcgruer commented:
> Aha! I knew we had some pre-generated tests, but not ones that are commit-time generated....

Me neither! It's why it took me far too long to realize what was going on :D

stephenmcgruer commented:
Firefox stability timed out (expected for a large CL). Chrome stability had some flakes; might be pre-existing but I am a bit concerned because I can't find previous PRs for the same files that show the flake. Retrying the CI to see if the flake changes.

stephenmcgruer commented:
Ok, so on the second run for Chrome stability, the flaky trusted-types/ failures disappeared, and the xhr/ failure remains:

```
65:21.88 INFO ## Unstable results ##
65:21.88 INFO |                  Test                  |                       Subtest                        |          Results           | Messages |
65:21.88 INFO |----------------------------------------|------------------------------------------------------|----------------------------|----------|
65:21.88 INFO | `/xhr/setrequestheader-bogus-name.htm` | `setRequestHeader should throw with header name "".` | **Duplicate subtest name** |          |
```

Looking at that test, it looks like that is an already-existing problem, so I'm not too worried about that.

In terms of reported results, [Safari](https://wpt.fyi/results/?diff&filter=ADC&run_id=393610007&run_id=416810013) and [Firefox](https://wpt.fyi/results/?diff&filter=ADC&run_id=395470012&run_id=391720018) are clear, no diff. [Chrome](https://wpt.fyi/results/?diff&filter=ADC&run_id=397680014&run_id=406970013) is showing a diff:

```
  | css/ | 5676 / 5997 | 5572 / 5997 | 0 / -104 / 0
  | kv-storage/ | 26 / 26 | 25 / 26 | 0 / -1 / 0
  | webrtc/ | 195 / 247 | 187 / 247 | 0 / -8 / 0
  | websockets/ | 64 / 180 | 60 / 180 | 0 / -4 / 0
```

New failures are:

- [css/css-properties-values-api/register-property-syntax-parsing.html](https://wpt.fyi/results/css/css-properties-values-api/register-property-syntax-parsing.html?diff&filter=ADC&run_id=397680014&run_id=406970013) : many
- [css/css-properties-values-api/register-property.html](https://wpt.fyi/results/css/css-properties-values-api/register-property.html?diff&filter=ADC&run_id=397680014&run_id=406970013) : two
- [css/css-typed-om/stylevalue-subclasses/numeric-objects/*](https://wpt.fyi/results/css/css-typed-om/stylevalue-subclasses/numeric-objects?diff&filter=ADC&run_id=397680014&run_id=406970013) : multiple tests
- [kv-storage/interface.https.html](https://wpt.fyi/results/kv-storage/interface.https.html?diff&filter=ADC&run_id=397680014&run_id=406970013) : Same-realm brand checks
- [webrtc/RTCConfiguration-iceServers.html](https://wpt.fyi/results/webrtc/RTCConfiguration-iceServers.html?diff&filter=ADC&run_id=397680014&run_id=406970013) : multiple, regarding SyntaxError
- [websockets/stream-tentative/*](https://wpt.fyi/results/websockets/stream-tentative?diff&filter=ADC&run_id=397680014&run_id=406970013) :  constructing with an invalid URL should throw

These all look valid and I've reproduced at least one locally. It's possible that these tests would work as expected if all of the changes in the original PR were applied together. For now, I am **reverting** the changes to those files, and we can look at adding them back later.

stephenmcgruer commented:
Ok, updated the commit with the following files reverted:

```
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   css/css-properties-values-api/register-property-syntax-parsing.html
	modified:   css/css-properties-values-api/register-property.html
	modified:   css/css-typed-om/stylevalue-subclasses/numeric-objects/cssMathValue.tentative.html
	modified:   css/css-typed-om/stylevalue-subclasses/numeric-objects/parse.tentative.html
	modified:   css/css-typed-om/stylevalue-subclasses/numeric-objects/to.tentative.html
	modified:   css/css-typed-om/stylevalue-subclasses/numeric-objects/toSum.tentative.html
	modified:   kv-storage/interface.https.html
	modified:   webrtc/RTCConfiguration-iceServers.html
	modified:   websockets/stream-tentative/constructor.any.js
```

Lets see how this does through the CI :)

bzbarsky commented:
> It's possible that these tests would work as expected if all of the changes in the original PR were applied together.

No, that shouldn't matter.

Looking at those tests,:

1) `css/css-properties-values-api/register-property-syntax-parsing.html` -- this should have `assert_throws_dom("SyntaxError", ...)`.  It was using the wrong thing before, and since this test is generally failing in Firefox I didn't catch this part.
2) `css/css-properties-values-api/register-property.html` -- same thing.  I should probably have looked for `new SyntaxError()` explicitly and fixed those up separately....
3) `css/css-typed-om/stylevalue-subclasses/numeric-objects/cssMathValue.tentative.html` -- same thing.
4) `css/css-typed-om/stylevalue-subclasses/numeric-objects/parse.tentative.html` -- same thing.
5) `css/css-typed-om/stylevalue-subclasses/numeric-objects/to.tentative.html` -- same thing.
6) `css/css-typed-om/stylevalue-subclasses/numeric-objects/toSum.tentative.html` -- same thing.
7)  `kv-storage/interface.https.html` -- the "Same-realm brand checks" test should be using `iframe.contentWindow.TypeError` throughout.
8)  `webrtc/RTCConfiguration-iceServers.html`-- this is the `SyntaxError` thing.
9) `websockets/stream-tentative/constructor.any.js` -- this is the `SyntaxError` thing.

Let's see how the rest of this stuff goes, and then I can do a separate PR to fix these tests properly.  Thank you again for sorting through this!

stephenmcgruer commented:
Stability checks fail as expected, but results are good for all browsers (no diffs). Merging (this was reviewed in https://github.com/web-platform-tests/wpt/pull/21350)

bzbarsky commented:
> and then I can do a separate PR to fix these tests properly

https://github.com/web-platform-tests/wpt/pull/21380
